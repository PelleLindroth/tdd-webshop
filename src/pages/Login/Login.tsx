import { useState } from 'react'
import { usersDb, IUser } from '../../mocks/users'

interface LoginProps {
  setUser: React.Dispatch<React.SetStateAction<IUser | null>>
}

const Login = (props: LoginProps) => {
  const { setUser } = props
  // const [email, setEmail] = useState('')

  return (
    <>
      <h1>LOGIN</h1>
      <form name="login-form">
        <label htmlFor="email">Email</label>
        <input name="email" id="email" type="text" />
        <label htmlFor="password">Password</label>
        <input name="password" id="password" type="text" />
        <button>Login</button>
      </form>
    </>
  )
}

export default Login
