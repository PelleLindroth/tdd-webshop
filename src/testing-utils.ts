import { render } from '@testing-library/react'
import { ReactElement } from 'react'
import { MemoryRouter } from 'react-router-dom'

const renderWithRouter = (ui: ReactElement, options?: any) => render(ui, { wrapper: MemoryRouter, ...options })

export { renderWithRouter }
