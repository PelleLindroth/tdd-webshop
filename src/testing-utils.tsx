import { render } from '@testing-library/react'
import { ReactElement } from 'react'
import { MemoryRouter } from 'react-router-dom'
import { Routes, Route } from 'react-router'

const renderWithRouter = (ui: ReactElement, options?: any) =>
  render(ui, { wrapper: MemoryRouter, ...options })

const withPath = (entry: string, element: ReactElement, path: string) => (
  <MemoryRouter initialEntries={[entry]}>
    <Routes>
      <Route path={path} element={element} />
    </Routes>
  </MemoryRouter>
)
export { renderWithRouter, withPath }
