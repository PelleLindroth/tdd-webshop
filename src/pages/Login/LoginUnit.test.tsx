import { shallow } from 'enzyme'
import { screen } from '@testing-library/react'
import { renderWithRouter } from '../../testing-utils'
import Login from './Login'
import userEvent from '@testing-library/user-event'

describe('Login unit tests', () => {
  it('renders Login component correctly', () => {
    const wrapper = shallow(<Login setUser={jest.fn()} />)

    expect(wrapper).toMatchSnapshot()
  })
  it('renders login form with text inputs for user email and password and a login button', () => {
    renderWithRouter(<Login setUser={jest.fn()} />)

    const emailInput = screen.getByRole('textbox', { name: /email/i })
    const passwordInput = screen.getByRole('textbox', { name: /password/i })
    const loginButton = screen.getByRole('button', { name: /login/i })

    expect(emailInput).toBeInTheDocument()
    expect(passwordInput).toBeInTheDocument()
    expect(loginButton).toBeInTheDocument()
  })
  it('updates values in email input when typing', () => {
    renderWithRouter(<Login setUser={jest.fn()} />)

    const emailInput = screen.getByRole('textbox', { name: /email/i })

    userEvent.type(emailInput, 'pelle')

    expect(screen.getByRole('form')).toHaveFormValues({
      email: 'pelle',
      password: '',
    })
  })
  it('updates values in password input when typing', () => {
    renderWithRouter(<Login setUser={jest.fn()} />)

    const passwordInput = screen.getByRole('textbox', { name: /password/i })

    userEvent.type(passwordInput, 'bananpaj')

    expect(screen.getByRole('form')).toHaveFormValues({
      email: '',
      password: 'bananpaj',
    })
  })
})

// updates values in text inputs when typing
// shows error message when submitting wrong credentials
// shows error message if either email or password input fields is empty
// renders my profile page correctly if user is logged in
// redirects to login from my profile if user is not logged in
