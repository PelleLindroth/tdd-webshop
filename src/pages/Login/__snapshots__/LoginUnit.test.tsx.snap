// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login unit tests renders Login component correctly 1`] = `
<Fragment>
  <h1>
    LOGIN
  </h1>
  <form
    name="login-form"
  >
    <label
      htmlFor="email"
    >
      Email
    </label>
    <input
      id="email"
      name="email"
      type="text"
    />
    <label
      htmlFor="password"
    >
      Password
    </label>
    <input
      id="password"
      name="password"
      type="text"
    />
    <button>
      Login
    </button>
  </form>
</Fragment>
`;
